devito/operations/interpolators.py

```diff
@@ -268,6 +268,15 @@ def inject(self, field, expr, offset=0):
         offset : int, optional
             Additional offset from the boundary.
         """
+
+        def get_field_offset(field):
```

Comment:
Could be done recursively, but I highly ever doubt this will be used. Not to mention that even for more than one level of nesting depth the indirections get really complex for more than one or two dimensions.

Reply:
perhaps you need a SparseFunction subclass overriding `origin`? just spitballing ideas, I don't even remember what `origin` is supposed to return

(@mloubout )

Reply:
`origin` is used for staggered functions interp/inject since they can have a half grid shift. Origin is the shift

Reply:
The recursion should be fixed now. Tests added.

Reply:
I'm really not happy... what is this function supposed to do exactly? 

also, if you have two indirections this function prefers the first one... ?

```
a[b(), c()]
```

will return `b.origin` ?

as I said, maybe you should just subclass Function or SparseFunction to your needs e.g. overriding `def origin` s.t.` it returns what you expect it to return? 

this really looks like a hack for whatever you're attempting to achieve, but happy to be proven wrong

@mloubout thoughts?

Reply:
I honestly don't think this makes much sense to me. For a given (staggered) function it has an origin for each of its indices. But I don't see how/why anyone would index a staggered function with "staggered indices", i.e what this does is something like

$ f(x + h_x, g(y + h_y) + h_y)$ which doesn't quite make any sense to me. It should be either $ f(x + h_x, g(y + h_y))$  or $ f(x + h_x, g(y) + h_y)$ to make sense IMO.

So as @FabioLuporini said and actual example would be useful on where such an odd case would show up.

Reply:
The main problem derives from the field_offset for `save_src[time, s(x, y, z)]` is (0,0) instead of (0,0,0).
This, a bit later in the code leads to no-generation of `pz`: 
```
    def _coordinate_bases(self, field_offset):
        """Symbol for the base coordinates of the reference grid point."""
        return tuple([cast_mapper[self.dtype](c - o - idx * i.spacing)
                      for c, o, idx, i, of in zip(self._coordinate_symbols,
                                                  self.grid.origin_symbols,
                                                  self._coordinate_indices,
                                                  self.grid.dimensions[:self.grid.dim],
                                                  field_offset)])
```

In all existing cases in master the field has the same space dimensions as self.grid

gird is also used for:

```
    @cached_property
    def _coordinate_indices(self):
        """Symbol for each grid index according to the coordinates."""
        return tuple([INT(FLOOR((c - o) / i.spacing))
                      for c, o, i in zip(self._coordinate_symbols,
                                         self.grid.origin_symbols,
                                         self.grid.dimensions
```

What about in case of not equal len of offset and grid dimensions to resort in grid dimensions always?
Thoughts?

Reply:
> The main problem derives from the field_offset for save_src[time, s(x, y, z)] is (0,0) instead of (0,0,0).

I get it, but as asked earlier, could you point us to the line of code that triggers the error without your edits? I guess it must be something along the lines of

```
sf.inject(...)
OR
sf.interpolate(...)
```